pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go

```diff
@@ -196,98 +200,196 @@ func (m *ManifestWorkController) sync(ctx context.Context, controllerContext fac
+		return nil, fmt.Errorf("Failed to decode object: %w", err)
 	}
-	mapping, err := m.restMapper.RESTMapping(unstructuredObj.GroupVersionKind().GroupKind(), unstructuredObj.GroupVersionKind().Version)
+
 	if err != nil {
-		return schema.GroupVersionResource{}, nil, fmt.Errorf("Failed to find gvr from restmapping: %w", err)
+		return nil, fmt.Errorf("Failed to find gvr from restmapping: %w", err)
 	}
 
-	return mapping.Resource, unstructuredObj, nil
+	return unstructuredObj, nil
 }
 
-func (m *ManifestWorkController) applyUnstructrued(ctx context.Context, data []byte, owner metav1.OwnerReference, recorder events.Recorder) (*unstructured.Unstructured, bool, error) {
-	gvr, required, err := m.decodeUnstructured(data)
+func (m *ManifestWorkController) applyUnstructrued(
+	ctx context.Context,
+	data []byte,
+	owner metav1.OwnerReference,
+	deleteOption *workapiv1.DeleteOption,
+	gvr schema.GroupVersionResource,
+	recorder events.Recorder) (*unstructured.Unstructured, bool, error) {
+
+	required, err := m.decodeUnstructured(data)
 	if err != nil {
 		return nil, false, err
 	}
 
-	required.SetOwnerReferences([]metav1.OwnerReference{owner})
 	existing, err := m.spokeDynamicClient.
 		Resource(gvr).
 		Namespace(required.GetNamespace()).
 		Get(ctx, required.GetName(), metav1.GetOptions{})
-	if errors.IsNotFound(err) {
+
+	switch {
+	case errors.IsNotFound(err):
 		actual, err := m.spokeDynamicClient.Resource(gvr).Namespace(required.GetNamespace()).Create(
-			ctx, required, metav1.CreateOptions{})
+			ctx, resourcemerge.WithCleanLabelsAndAnnotations(required).(*unstructured.Unstructured), metav1.CreateOptions{})
 		recorder.Eventf(fmt.Sprintf(
 			"%s Created", required.GetKind()), "Created %s/%s because it was missing", required.GetNamespace(), required.GetName())
 		return actual, true, err
-	}
-	if err != nil {
+	case err != nil:
 		return nil, false, err
 	}
 
+	// Merge OwnerRefs.
+	required.SetOwnerReferences([]metav1.OwnerReference{owner})
+	existingOwners := existing.GetOwnerReferences()
+	modified := resourcemerge.BoolPtr(false)
+
+	resourcemerge.MergeOwnerRefs(modified, &existingOwners, required.GetOwnerReferences())
+
+	if *modified {
+		required.SetOwnerReferences(existingOwners)
+	}
+
 	// Compare and update the unstrcuctured.
 	if isSameUnstructured(required, existing) {
 		return existing, false, nil
 	}
 	required.SetResourceVersion(existing.GetResourceVersion())
 	actual, err := m.spokeDynamicClient.Resource(gvr).Namespace(required.GetNamespace()).Update(
-		context.TODO(), required, metav1.UpdateOptions{})
+		ctx, required, metav1.UpdateOptions{})
 	recorder.Eventf(fmt.Sprintf(
 		"%s Updated", required.GetKind()), "Updated %s/%s", required.GetNamespace(), required.GetName())
 	return actual, true, err
 }
 
+// manageOwnerRef update ownerref based on the resource and the deleteOption. If the resource is orphaned.
+// The owner's UID is updated to indicate that the owner should be removed from the existing ownerrefs in the resource.
```

Comment:
it actually returns a reference indicating that the owner should be removed, right?

Reply:
yes. I will update the comment.